### ISSUE TITLE
`keyword-spacing` incorrectly reports `else ++foo`

### ISSUE THREAD

**Issue by Alhadis** [eslint/eslint#12778]
### Environment
* **ESLint Version:** `v6.8.0`
* **Node Version:** `v13.6.0`
* **npm Version:** `v6.13.4`
* **Parser:** None (default)

### Steps to reproduce:

1.	Create a file named `test.js` containing:
	~~~js
	var good = 0;
	var bad = 0;

	if(true){
		++good;
	}
	else ++bad;
	~~~

2.	Create an `.eslintrc.json` config containing:
	~~~json
	{
		"rules": {
			"keyword-spacing": [
				"error",
				{
					"overrides": {
						"if":      {"after": false},
						"else":    {"after": false},
						"for":     {"after": false},
						"while":   {"after": false},
						"switch":  {"after": false},
						"try":     {"after": false},
						"catch":   {"after": false},
						"finally": {"after": false},
						"with":    {"after": false}
					}
				}
			]
		}
	}
	~~~

3.	Run `eslint test.js`.

### Expectation

`else ++bad` shouldn't have triggered an error (note that the error vanishes if I use `else bad++` instead).

### Reality

~~~console
$ eslint test.js

/Users/john/Desktop/tmp/test.js
  9:1  error  Unexpected space(s) after "else"  keyword-spacing

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
~~~

### Are you willing to submit a pull request to fix this bug?

Yeah.

**Comment by mdjermanovic** [comment]
Hi @Alhadis, thanks for the bug report!

The rule is configured to disallow unnecessary spaces after `else`, so it expects `else++bad` which is a valid code (unlike `elsebad++`).

This seems to be intended behavior, there is a somewhat similar [test case with else+1](https://github.com/eslint/eslint/blob/a1d999c6b4e51c317ad409110be7be214ff9f7c6/tests/lib/rules/keyword-spacing.js#L1957)

**Comment by Alhadis** [comment]
@mdjermanovic Do you really think users who configure `keyword-spacing` to disallow spaces in `else (` intended to mean `else++bad`…?

If it *is* intentional (and I seriously hope it isn't), there needs to be an option to accommodate this.

**Comment by mdjermanovic** [comment]
I agree that `else++bad` doesn't look good.

I hope someone from the team can provide more context on this behavior (as it might be intentional) and should we treat this as a bug or an enhancement request.

**Comment by kaicataldo** [comment]
It looks like this is working as expected to me. I think it could make sense to add some option that differentiates between keywords followed by blocks (`else { ++counter; }`) vs keywords that are follows by statements (`else ++counter;`), but we'll need to figure out an API that works well with the current options.

**Comment by Alhadis** [comment]
```js
"before": true | false | "blocks-only" | "statements-only"
```

**Comment by Alhadis** [comment]
> but we'll need to figure out an API that works well with the current options.

@kaicataldo What about:

```js
["before" | "after" | keyword]: true | false | "only-blocks" | "only-statements" | "never-blocks" | "never-statements"
```

... where `"only-"` and `"never"-` correspond to `true` and `false`, except that only statements or blocks are affected.

**Comment by Alhadis** [comment]
@kaicataldo, @mdjermanovic Is there any action required on my part?

**Comment by Alhadis** [comment]
Is this issue going to get fixed? Or will it just get flagged as stale and closed like my last two attempts at improving your software?

**Comment by kaicataldo** [comment]
@Alhadis Please keep in my mind that most of the team is volunteering their free time to work on the project, and we therefore have to prioritize pretty ruthlessly to make sure we can maintain the project.

Regarding the API, I wonder if something alone the lines of the following would work:
```ts
/*
eslint keyword-spacing: [
    "error", { 
        "before": boolean
        "after": boolean, 
        "beforeBlock": { "before": boolean, "after": boolean },
        "beforeStatement": { "before": boolean, "after": boolean },
        "overrides": {
            [string]: { "before": boolean, "after": boolean }
        }
    }
] 
*/
```

**Comment by Alhadis** [comment]
Thanks. Should I submit a pull-request?

**Comment by kaicataldo** [comment]
@Alhadis You're welcome to, though we generally advise for folks to hold off until the team has reached consensus on an issue to avoid throwaway work.

Alternatively, since ESLint is fully pluggable, you can also create your own custom rule.

@eslint/eslint-team Thoughts on this?

**Comment by not-an-aardvark** [comment]
To me, this seems like it's working as intended and it's not clear that it's worth adding a new option for (we usually encourage people to create custom rules in cases like this if their preferences don't match the built-in rule). But I don't feel strongly about it, so I'll defer to the consensus if other team members disagree with me.

**Comment by Alhadis** [comment]
> we usually encourage people to create custom rules in cases like this if their preferences don't match the built-in rule

@not-an-aardvark And what would such a plugin look like?

~~~shell
npm install -D eslint-plugin-keyword-spacing-but-with-space-before-statements
~~~

I'd rather not publish and maintain a piece of software that does exactly what ESLint does, but with one minor exception.

**Comment by not-an-aardvark** [comment]
That's understandable, but it's also not realistic for the ESLint team to support every user's style preferences, so we typically only add support for options that would benefit a large percentage of users. (My opinion is that this isn't one of those cases, although it seems like there is room for disagreement on that point.) If you did end up creating a plugin or project-specific rule, you could use something like [`eslint-rule-composer`](https://github.com/not-an-aardvark/eslint-rule-composer) to make the implementation quite small.

**Comment by Alhadis** [comment]
It's a stretch to consider this a matter of style or preference instead of an inconsistency with ESLint. `keyword-spacing` permits `else foo++` but forbids `else ++foo` simply because the latter *can* be written without a space. That's different to something like `if (…)` vs. `if(…)`, where the omission of whitespace has no bearing on how the code gets interpreted.

> you could use something like `eslint-rule-composer` to make the implementation quite small.

Yes, that was recommended to me the last two times I filed an issue here (including once I'm *certain* [qualified as a bug](https://github.com/eslint/eslint/issues/10726)). Hence why I half-expected yet another dismissal when opening this issue.

**Comment by kaicataldo** [comment]
@Alhadis The issue linked looks like it was closed because of a lack of discussion after we responded.

I agree that this is working as expected (given the current iteration of the rule), but also understand why this style would not be desirable. I would support an enhancement if we can reach consensus around what a reasonable API would be.

Just one more reminder that ESLint is maintained mostly by volunteers who work on the project in their spare time and, given the resources we have, we have to prioritize what we work on heavily. This means that we can’t accept every enhancement proposal (even ones that are reasonable), as much as we would like to. ESLint is pluggable for this reason, though, so that users can enforce whatever code style as they like, whether we have the bandwidth to implement it in core rules or not.

The next step here is to try to reach consensus with other team members, since @not-an-aardvark mentioned that he’ll defer to the general consensus of the rest of the team.

**Comment by Alhadis** [comment]
I understand you're only human and do this in your own free-time. That's why I'm all too happy to help.

If you're that overburdened, maybe I shouldn't report here anymore.

**Comment by Alhadis** [comment]
Done here. You won't see me here ever again.
